### PR TITLE
Decode model schema name on model component

### LINKED ISSUE #5626: Path in $ref is partually percent-encoded
### Q&A
 - OS: windows
 - Browser: chrome
 - Version: 76.0.3809.132
 - Method of installation: no installation, issue is presented at editor.swagger.io
 - Swagger-Editor version: current editor.swagger.io
 - Swagger/OpenAPI version: Swagger 2.0

### Content & configuration

Swagger/OpenAPI definition:
```
swagger:` '2.0'
info:
  description: 'Swagger docs'
  title: API example
  version: '1.0'
host: server
basePath: /docs
paths:
  '/ws/person':
    get:
      parameters:
        - name: id
          type: number
          in: path
          description: 'person id'
          required: true
      description: Get person
      operationId: 'get_person'
      tags:
        - Person
      responses:
        '200':
          description: Ok
          schema:
            type: object
            properties:
              result:
                $ref: '#/definitions/custom%3A%3Anamespace%3A%3APerson'
definitions:
  'custom::namespace::Person':
    properties:
      id:
        type: number
      FirstName:
        type: string
      LastName:
        type: string
      Address:
        type: array
        items:
          $ref: '#/definitions/custom%3A%3Anamespace%3A%3AAddress'
  'custom::namespace::Address':
    properties:
      id:
        type: number
      street:
        type: string
      city:
        type: string
```

### How can we help?

In order to generate JSON without warnings some part $ref needs to be percent encoded, and not. Is this correct behavior, or maybe bug.

Way to reproduce this issue:
* Current JSON that contains part without and part with percent encoding will not have warnings.
* If we convert $ref without percent encoding warning will popup `$ref: '#/definitions/custom::namespace::Person'`
* If we convert $ref with all characters percent encoding warning will popup `$ref: '%23%2Fdefinitions%2Fcustom%3A%3Anamespace%3A%3APerson'`

### PR DESCRIPTION
**Pull request by souenzzo**
### Description
Decodes the schema name on model 

### Motivation and Context
Fixes #5626 (for both OpenAPI v3 and Swagger V2)

### How Has This Been Tested?
- Go to `localhost:3200`
- Paste this data-url in  the url field

```
data:application/json;base64,eyJjb21wb25lbnRzIjp7InNjaGVtYXMiOnsiYVwvYitjIjp7InR5cGUiOiJudW1iZXIifX19LCJvcGVuYXBpIjoiMy4wLjMiLCJwYXRocyI6eyJcLyI6eyJnZXQiOnsicmVzcG9uc2VzIjp7ImRlZmF1bHQiOnsiY29udGVudCI6eyJhcHBsaWNhdGlvblwvanNvbiI6eyJzY2hlbWEiOnsiJHJlZiI6IiNcL2NvbXBvbmVudHNcL3NjaGVtYXNcL2F+MWIlMkJjIn19fX19fX19fQ==
```

- Expand `/` route
- Switch the default response from `example value` to `model`
- You should see `a/b+c`, without any `~1` or `%B2`

An data-uri to test Swagger v2 spec

```
data:application/yaml;base64,c3dhZ2dlcjogJzIuMCcKaW5mbzoKICBkZXNjcmlwdGlvbjogJ1N3YWdnZXIgZG9jcycKICB0aXRsZTogQVBJIGV4YW1wbGUKICB2ZXJzaW9uOiAnMS4wJwpob3N0OiBzZXJ2ZXIKYmFzZVBhdGg6IC9kb2NzCnBhdGhzOgogICcvd3MvcGVyc29uJzoKICAgIGdldDoKICAgICAgZGVzY3JpcHRpb246IEdldCBwZXJzb24KICAgICAgb3BlcmF0aW9uSWQ6ICdnZXRfcGVyc29uJwogICAgICB0YWdzOgogICAgICAgIC0gUGVyc29uCiAgICAgIHJlc3BvbnNlczoKICAgICAgICAnMjAwJzoKICAgICAgICAgIGRlc2NyaXB0aW9uOiBPawogICAgICAgICAgc2NoZW1hOgogICAgICAgICAgICB0eXBlOiBvYmplY3QKICAgICAgICAgICAgcHJvcGVydGllczoKICAgICAgICAgICAgICByZXN1bHQ6CiAgICAgICAgICAgICAgICAkcmVmOiAnIy9kZWZpbml0aW9ucy9jdXN0b20lM0ElM0FuYW1lc3BhY2UlM0ElM0FQZXJzb24nCmRlZmluaXRpb25zOgogICdjdXN0b206Om5hbWVzcGFjZTo6UGVyc29uJzoKICAgIHByb3BlcnRpZXM6CiAgICAgIGlkOgogICAgICAgIHR5cGU6IG51bWJlcgogICAgICBGaXJzdE5hbWU6CiAgICAgICAgdHlwZTogc3RyaW5nCiAgICAgIExhc3ROYW1lOgogICAgICAgIHR5cGU6IHN0cmluZwogICAgICBBZGRyZXNzOgogICAgICAgIHR5cGU6IGFycmF5CiAgICAgICAgaXRlbXM6CiAgICAgICAgICAkcmVmOiAnIy9kZWZpbml0aW9ucy9jdXN0b20lM0ElM0FuYW1lc3BhY2UlM0ElM0FBZGRyZXNzJwogICdjdXN0b206Om5hbWVzcGFjZTo6QWRkcmVzcyc6CiAgICBwcm9wZXJ0aWVzOgogICAgICBpZDoKICAgICAgICB0eXBlOiBudW1iZXIKICAgICAgc3RyZWV0OgogICAgICAgIHR5cGU6IHN0cmluZwogICAgICBjaXR5OgogICAgICAgIHR5cGU6IHN0cmluZwogICAgICAgIA==
```

### Screenshots (if appropriate):
Screenshot 2023-02-21 at 10 15 54
Before screenshot in  #5626

### My PR contains... 
- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [x] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by souenzzo** [comment]
@ponelat

**Comment by ponelat** [comment]
Thanks for the fix @souenzzo.
When you say the test won't run... does it run? I don't understand. 
Happy to get this merged, as the model name is mostly visual. Not sure about all the edge-cases of JSON Pointer and URL encoding. 

@souenzzo could you update the branch. 
@char0n any reason the PR builders aren't running, is it to do with the Draft Status of this PR?

File: src/core/components/model.jsx [hunk 1]
```diff
@@ -3,6 +3,8 @@ import ImmutablePureComponent from "react-immutable-pure-component"
 import ImPropTypes from "react-immutable-proptypes"
 import PropTypes from "prop-types"
 
+export let decodeRefName = s => decodeURIComponent(s.replace(/~1/, "/").replace(/~0/, "~"))
```
Comment (char0n) [inline src/core/components/model.jsx:6]: Again, thanks for contributing!

**1.** [decodeURIComponent](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent) can throw [URIError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/URIError) in certain cases like for example: `c%d`. JSON Pointer specification unfortunately allow to use encoded strings like  `c%d`, which means we have to compensate for that cornser-case. More info in [RFC6901, section 5](https://www.rfc-editor.org/rfc/rfc6901#section-5).

**2.** The regex pattern you used only replaces first occurrence of encoded JSON Pointer symbols and ignore the rest.

Here is a suggestion for you function that addressed both the issues mentioned above:

```suggestion
export let decodeRefName = uri => {
  const unescaped = uri.replace(/~1/g, "/").replace(/~0/g, "~")
  
  try {
    return decodeURIComponent(unescaped)
  } catch {
    return unescaped
  }
}
```

PS: can I you pleas also provide more tests coverage for the cases that my suggestion tries to fix? Thanks!

**Review by char0n — CHANGES_REQUESTED** [review]
Please have a look at my code review comments.

**Comment by souenzzo** [comment]
@char0n done

**Review by char0n — APPROVED** [review]
LGTM

File: src/core/components/model.jsx [hunk 2]
```diff
@@ -3,6 +3,16 @@ import ImmutablePureComponent from "react-immutable-pure-component"
 import ImPropTypes from "react-immutable-proptypes"
 import PropTypes from "prop-types"
 
+let decodeRefName = uri => {
```
Comment (char0n) [inline src/core/components/model.jsx:6]:
```suggestion
const decodeRefName = uri => {
```

Comment (char0n) [reply src/core/components/model.jsx:6]: We broke away from the `let` some time ago, and we use now industry strandard practices and use `const` whenever possible.

**Comment by char0n** [comment]
Did some final style changes before the merge.

**Comment by souenzzo** [comment]
@char0n tnks :) 

Btw, a feedback from the contribution process:

would be nice if the linter warn us about const vs let usage.

**Comment by char0n** [comment]
Hi @souenzzo,

> would be nice if the linter warn us about const vs let usage.

Yes, automated linting is the plan. Utilizing prettier + eslint + aribnb codestyle will be utilized (as it's already has been done for SwaggerEditor and SwaggerClient). That would require to convert entire codebase. Meanwhile we have to check that manually unfortunately which decrease dev experience. Thanks for the feedback!
